src/freenet/clients/fcp/FCPPluginConnectionImpl.java

```diff
@@ -0,0 +1,1254 @@
+     *  Used as performance optimization to prevent construction of the log strings if it is not
+     *  necessary. */
+    private static transient volatile boolean logMINOR = false;
+    
+    static {
+        // Necessary for automatic setting of logDEBUG and logMINOR
+        Logger.registerClass(FCPPluginConnectionImpl.class);
+    }
+
+
+    /**
+     * Unique identifier among all {@link FCPPluginConnection}s and FCPPluginConnectionImpls
+     * @see #getID()
+     */
+    private final UUID id = UUID.randomUUID();
+
+    /**
+     * Executor upon which we run threads of the send functions.<br>
+     * Since the send functions can be called very often, it would be inefficient to create a new
+     * {@link Thread} for each one. An {@link Executor} prevents this by having a pool of Threads
+     * which will be recycled.
+     */
+    private final Executor executor;
+
+    /**
+     * The class name of the plugin to which this FCPPluginConnectionImpl is connected.
+     */
+    private final String serverPluginName;
+
+    /**
+     * The FCP server plugin to which this connection is connected.
+     * 
+     * <p>TODO: Optimization / Memory leak fix: Monitor this with a {@link ReferenceQueue} and if it
+     * becomes nulled, remove this FCPPluginConnectionImpl from the map
+     * {@link FCPConnectionHandler#pluginConnectionsByServerName}.<br/>
+     * Currently, it seems not necessary:<br/>
+     * - It can only become null if the server plugin is unloaded / reloaded. Plugin unloading /
+     *   reloading requires user interaction or auto update and shouldn't happen frequently.<br/>
+     * - It would only leak one WeakReference per plugin per client network connection. That won't
+     *   be much until we have very many network connections. The memory usage of having one thread
+     *   per {@link FCPConnectionHandler} to monitor the ReferenceQueue would probably outweigh the
+     *   savings.<br/>
+     * - We already opportunistically clean the table at FCPConnectionHandler: If the client
+     *   application which is behind the {@link FCPConnectionHandler} tries to send a message using
+     *   a FCPPluginConnectionImpl whose server WeakReference is null, it is purged from the said
+     *   table at FCPConnectionHandler. So memory will not leak as long as the clients keep trying
+     *   to send messages to the nulled server plugin - which they probably will do because they did
+     *   already in the past.<br/>
+     * NOTICE: If you do implement this, make sure to not rewrite the ReferenceQueue polling thread
+     *         but instead base it upon {@link FCPPluginConnectionTracker}. You should probably
+     *         extract a generic class WeakValueMap from that one and use to to power both the
+     *         existing class and the one which deals with this variable here.
+     * </p>
+     * @see #isServerDead() Use isServerDead() to check whether this WeakReference is nulled.
```

Comment:
Why is there a function to check whether a reference is null? What abstraction is it intended to gain? Unless I'm missing something if one uses isServerDead(), they call it first, then get the reference, then have to check if it's null anyway because it's not impossible for it to have been GC'd in the meantime.